src/anyp/PortCfg.cc

```diff
@@ -101,3 +105,185 @@ AnyP::PortCfg::detailCodeContext(std::ostream &os) const
+    }
+
+    debugs(3, DBG_CRITICAL, "FATAL: " << PortOptionStr(aFlag) << " requires one of " << PortOptionStrList(list) <<
+            " on " << PortKindStr(rawFlags) << ' ' << detail);
+    self_destruct();
+}
+
+void
+AnyP::PortCfg::checkImplication(const AnyP::TrafficModeFlags::Pointer aFlag, const AnyP::TrafficModeFlags::Pointer otherFlag, const char *detail)
+{
+    const auto &rawFlags = flags.rawConfig();
+    if (!(rawFlags.*aFlag))
+        return;
+    if (!(rawFlags.*otherFlag)) {
+        debugs(3, DBG_CRITICAL, "FATAL: " << PortOptionStr(aFlag) << " requires " << PortOptionStr(otherFlag) <<
+                " on " << PortKindStr(rawFlags) << ' ' << detail);
+        self_destruct();
+    }
+}
+
+bool
+AnyP::PortCfg::hasAll(const AnyP::TrafficModeFlags::List &list)
+{
+    const auto &rawFlags = flags.rawConfig();
+    for (const auto &p: list) {
+        if (!(rawFlags.*p))
+            return false;
+    }
+    return true;
+}
+
+void
+AnyP::PortCfg::checkFlags()
+{
+    using Flags = AnyP::TrafficModeFlags;
+    const auto &rawFlags = flags.rawConfig();
+
+    switch (rawFlags.portKind) {
+
+    case Flags::httpPort: {
+        if (rawFlags.accelSurrogate)
+            rejectFlags({&TrafficModeFlags::natIntercept, &TrafficModeFlags::tproxyIntercept}, "accel");
+        else
+            allowEither({&TrafficModeFlags::natIntercept, &TrafficModeFlags::tproxyIntercept});
+    }
+    break;
+
+    case Flags::httpsPort: {
+        if (rawFlags.accelSurrogate) {
+            rejectFlags({&TrafficModeFlags::natIntercept, &TrafficModeFlags::tproxyIntercept,
+                    &TrafficModeFlags::proxySurrogateHttp, &TrafficModeFlags::tunnelSslBumping}, "accel");
+        } else {
+            // ssl-bump requires one of: tproxy, intercept or require-proxy-header
+            checkImplication(&TrafficModeFlags::tunnelSslBumping,
+                    {&TrafficModeFlags::natIntercept, &TrafficModeFlags::tproxyIntercept, &TrafficModeFlags::proxySurrogateHttp});
+
+            // tproxy, intercept and require-proxy-header require ssl-bump
+            checkImplication(&TrafficModeFlags::natIntercept, &TrafficModeFlags::tunnelSslBumping);
+            checkImplication(&TrafficModeFlags::tproxyIntercept, &TrafficModeFlags::tunnelSslBumping);
+            checkImplication(&TrafficModeFlags::proxySurrogateHttp, &TrafficModeFlags::tunnelSslBumping);
+
+            allowEither({&TrafficModeFlags::natIntercept, &TrafficModeFlags::tproxyIntercept, &TrafficModeFlags::proxySurrogateHttp});
+        }
+    }
+    break;
+
+    case Flags::ftpPort:
+        rejectFlags({&TrafficModeFlags::accelSurrogate, &TrafficModeFlags::proxySurrogateHttp, &TrafficModeFlags::tunnelSslBumping});
+        allowEither({&TrafficModeFlags::natIntercept, &TrafficModeFlags::tproxyIntercept});
+        break;
+
+    default:
+        fatal("unreachable");
+    }
+
+    if (hasAll({&TrafficModeFlags::tproxyIntercept, &TrafficModeFlags::proxySurrogateHttp})) {
```

Comment:
I suggest removing hasAll() because an explicit list of and-ed flags in this `if` condition would look simpler.